### PR TITLE
Navigation block e2e tests: default to a list of pages if there are no menus

### PR DESCRIPTION
**Pull request by MaggieCabrera**
## What?

This is part of the effort to re work all the navigation block e2e tests and migration to playwright. This is one of the user stories described in #45199

## Why?

The navigation e2e tests were really flaky and bottlenecking the github actions. We are rethinking them so they make more sense and are more usable and maintainable.

## Testing Instructions

Run `npm run test:e2e:playwright -- editor/blocks/navigation.spec.js`

Co-authored-by: Dave Smith <444434+getdave@users.noreply.github.com>

### CODE REVIEW

File: test/e2e/specs/editor/blocks/navigation.spec.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+			const pageListSelector = await editor.page.locator(
+				`//table[contains(@aria-label,'Block navigation structure')]//a[.//span[text()='Page List']]`
+			);
+
+			expect( pageListSelector ).toBeTruthy();
+
+			// Check the markup of the block is correct.
+			await editor.publishPost();
+			const content = await editor.getEditedPostContent();
+			expect( content ).toBe(
```
Comment (draganescu) [inline test/e2e/specs/editor/blocks/navigation.spec.js:55]: Nit: I think this should be - I know it's very granular - a separate battery of functional tests about page list in navigation where we also test for instance that the fallback works on the front end.

**Comment by MaggieCabrera** [comment]
Looks like this test needs work since the page list block has changed :D

File: test/e2e/specs/editor/blocks/navigation.spec.js (hunk 1 above)
Comment (MaggieCabrera) [reply test/e2e/specs/editor/blocks/navigation.spec.js:55]: Should we add those to the list in https://github.com/WordPress/gutenberg/issues/44920 for a follow up?

File: test/e2e/specs/editor/blocks/navigation.spec.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+		test.beforeAll( async ( { requestUtils } ) => {
+			await requestUtils.activateTheme( 'emptytheme' );
+		} );
+
+		test.beforeEach( async ( { admin } ) => {
+			await admin.createNewPost();
+		} );
+
+		test.afterAll( async ( { requestUtils } ) => {
+			await requestUtils.activateTheme( 'twentytwentyone' );
```
Comment (getdave) [inline test/e2e/specs/editor/blocks/navigation.spec.js:28]: Why is this necessary?

**Review by getdave — COMMENTED** [review]
Thanks for moving these forward 🎉 

I noticed that we're not actually checking that Pages are rendered because we don't wait for the request to complete. Hopefully that's ok as the Page List tests should be responsible for that I guess.

File: test/e2e/specs/editor/blocks/navigation.spec.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+				`//a[.//span[text()='Navigation']]/span[contains(@class, 'block-editor-list-view__expander')]`,
+				{ force: true }
+			);
+
+			// Find the Page list selector inside the navigation block.
+			const pageListSelector = await page.locator(
+				`//table[contains(@aria-label,'Block navigation structure')]//a[.//span[text()='Page List']]`
+			);
+
+			expect( pageListSelector ).toBeTruthy();
```
Comment (getdave) [inline test/e2e/specs/editor/blocks/navigation.spec.js]: I believe this will provide a false positive in the sense that the `locator` will always return an object which is therefore truthy.

What I think we want is `await expect( pageListSelector ).toBeVisible();` but this will currently fail the test becayse the request (the Nav block loading state) won't have resolved.

File: test/e2e/specs/editor/blocks/navigation.spec.js (hunk 1 above)
Comment (getdave) [reply test/e2e/specs/editor/blocks/navigation.spec.js:55]: @draganescu Are you saying the premise of this test is void? Can you clarify what you see as next steps?

File: test/e2e/specs/editor/blocks/navigation.spec.js (hunk 3 above)
Comment (getdave) [reply test/e2e/specs/editor/blocks/navigation.spec.js]: Fixed.

File: test/e2e/specs/editor/blocks/navigation.spec.js (hunk 1 above)
Comment (draganescu) [reply test/e2e/specs/editor/blocks/navigation.spec.js:55]: I think the premise is good, I don't remember what I meant :/

File: test/e2e/specs/editor/blocks/navigation.spec.js (hunk 2 above)
Comment (MaggieCabrera) [reply test/e2e/specs/editor/blocks/navigation.spec.js:28]: I was testing and my test env was loading TT1 by default, which is not a block theme, and it was making the test fail. I've seen other tests do this same thing and since we rely on having a block theme installed, it felt like the correct way to go about it? I'm not sure!

Comment (getdave) [reply test/e2e/specs/editor/blocks/navigation.spec.js:28]: Thanks. That's useful context.

Comment (scruffian) [reply test/e2e/specs/editor/blocks/navigation.spec.js:28]: We should probably explain that in a comment.

File: test/e2e/specs/editor/blocks/navigation.spec.js [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+test.describe(
+	'As a user I want the navigation block to fallback to the best possible default',
+	() => {
+		test.beforeAll( async ( { requestUtils } ) => {
+			await requestUtils.activateTheme( 'emptytheme' );
+		} );
+
+		test.beforeEach( async ( { admin } ) => {
+			await admin.createNewPost();
```
Comment (scruffian) [inline test/e2e/specs/editor/blocks/navigation.spec.js:24]: Is there any point in doing this if we then delete all posts?

File: test/e2e/specs/editor/blocks/navigation.spec.js [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+			// Check Page List is in the list view.
+
+			// Open the list view.
+			await page.click( '[aria-label="Document Overview"i]' );
+
+			const listView = await page.locator(
+				`//table[contains(@aria-label,'Block navigation structure')]`
+			);
+
+			// Click the Navigation block expander within the List View.
```
Comment (scruffian) [inline test/e2e/specs/editor/blocks/navigation.spec.js]: I don't think we need to test this. It's a test for the list view, not a test for the navigation block.

Comment (getdave) [reply test/e2e/specs/editor/blocks/navigation.spec.js]: The reason we do this is to have a generic way to check that page list is "visible" to the user.

Alternatives

- remove entirely and rely on non-user perceivable things like the snapshot of the code (see end of test)
- assert on presence of block(s) within the editor _canvas_

File: test/e2e/specs/editor/blocks/navigation.spec.js (hunk 4 above)
Comment (MaggieCabrera) [reply test/e2e/specs/editor/blocks/navigation.spec.js:24]: I've seen this done all over the tests for the rest of the blocks, I can't remember off the top of my head where I read it but I think it was good practice to clean up after the tests are done, so the environment stays how it started before testing. We need to create the post before we test the block inside it though, don't we? Or maybe I'm not understanding your concern.

Comment (getdave) [reply test/e2e/specs/editor/blocks/navigation.spec.js:24]: You need a fresh post to run the tests against. As is the cleanup. Otherwise this test could pollute other tests. That said I wonder whether this behavior is baked into the framework itself. Maybe @kevin940726 knows?

Comment (kevin940726) [reply test/e2e/specs/editor/blocks/navigation.spec.js:24]: We don't automatically delete all posts after each test, but we do delete them all once before starting all tests. And yeah, not every test needs a fresh post so we don't do that automatically either. I think putting this in `beforeEach` is a good practice 👍 .

File: test/e2e/specs/editor/blocks/navigation.spec.js (hunk 2 above)
Comment (kevin940726) [reply test/e2e/specs/editor/blocks/navigation.spec.js:28]: Yeah, due to WordPress being a stateful environment, we kind of have to reset the state after each/all tests.

File: test/e2e/specs/editor/blocks/navigation.spec.js [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+		test( 'default to a list of pages if there are no menus', async ( {
+			editor,
+			page,
+		} ) => {
+			await editor.insertBlock( { name: 'core/navigation' } );
+
+			// Necessary to wait for the block to be in loaded state else
+			// initial rendered state will not be resolved.
+			await page.waitForLoadState( 'networkidle' );
```
Comment (kevin940726) [inline test/e2e/specs/editor/blocks/navigation.spec.js]: This is normally not needed as the actions below will automatically (implicitly) wait for the result. Try removing it and see if it still works?

File: test/e2e/specs/editor/blocks/navigation.spec.js [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+			// Check Page List is in the list view.
+
+			// Open the list view.
+			await page.click( '[aria-label="Document Overview"i]' );
```
Comment (kevin940726) [inline test/e2e/specs/editor/blocks/navigation.spec.js]: Have you read the [best practices](https://github.com/WordPress/gutenberg/blob/trunk/test/e2e/README.md#use-accessible-selectors) guide? We should favor role selectors wherever possible, it's more readable, accessible, and free of implementation details. The same goes for all the other selectors below.

File: test/e2e/specs/editor/blocks/navigation.spec.js [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+			const listView = await page.locator(
```
Comment (kevin940726) [inline test/e2e/specs/editor/blocks/navigation.spec.js]: locator doesn't return a promise. It's a common misconception coming from Puppeteer 😅 .

File: test/e2e/specs/editor/blocks/navigation.spec.js (hunk 6 above)
Comment (getdave) [reply test/e2e/specs/editor/blocks/navigation.spec.js]: It doesn't. The Page List isn't available in the List View until the network resolves. I have been trying to find a way around this.

File: test/e2e/specs/editor/blocks/navigation.spec.js (hunk 7 above)
Comment (getdave) [reply test/e2e/specs/editor/blocks/navigation.spec.js]: I did infact read that guide. That said I was finding it hard to see how to use the `getByRole()` to select something with an `aria-label`. I will check again to see if I can figure it out.

File: test/e2e/specs/editor/blocks/navigation.spec.js (hunk 8 above)
Comment (getdave) [reply test/e2e/specs/editor/blocks/navigation.spec.js]: Thanks 👍

File: test/e2e/specs/editor/blocks/navigation.spec.js (hunk 7 above)
Comment (getdave) [reply test/e2e/specs/editor/blocks/navigation.spec.js]: Ah `getByLabelText` should work

**Comment by getdave** [comment]
@kevin940726 I've tried to wait on specific network requests. But every time I try running these tests the `isVisible` `expect` always timesout. 

I looked at the artifacts provided by the test runner and no matter what I do the Page List block only becomes visible in the editor after the `isVisible` check runs. I even tried adding a 10s timeout!

I'm a bit stuck at this point. Any help or advice would be much appreciated.

**Comment by kevin940726** [comment]
> But every time I try running these tests the isVisible expect always timesout.

Thanks for the ping! I believe it's because we didn't use `editor.canvas` to access the element. Since now the post editor is iframed, we have to modify our selector to select from the iframe canvas. In order to do that, we also have to initiate the `editor` utils with the `hasIframe` option set to true. It's a bit tedious now that almost every editor is iframed though. I started a separate discussion about that in https://github.com/WordPress/gutenberg/pull/40815#issuecomment-1374601661.

I hope it's okay for me to push the change directly on this branch :).

File: test/e2e/specs/editor/blocks/navigation.spec.js [hunk 9]
```diff
@@ -24,32 +34,17 @@ test.describe(
...
-			const menusRequest = page
-				.waitForRequest( ( response ) => {
-					return response.url().includes( 'rest_route=/wp/v2/menus' );
-				} )
-				.catch( () => {} );
-
 			await editor.insertBlock( { name: 'core/navigation' } );
 
-			await pagesRequest;
-			await menusRequest;
```
Comment (kevin940726) [inline test/e2e/specs/editor/blocks/navigation.spec.js]: We don't need to explicitly wait for these implementation details. Instead, the recommended way is to wait for what the users expect. In this case, we can just wait for the page list block to be visible.

Comment (getdave) [reply test/e2e/specs/editor/blocks/navigation.spec.js]: That would be 100% ideal. Let's hope we can do this. But isn't it potentially flaky to rely on network resolving faster than the Locator timeout value? Should I bump the timeout?

File: test/e2e/specs/editor/blocks/navigation.spec.js [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+		} ) => {
+			await editor.insertBlock( { name: 'core/navigation' } );
+
+			const pageListBlock = editor.canvas.getByRole( 'document', {
+				name: 'Block: Page List',
+			} );
+
+			await expect( pageListBlock ).toBeVisible( {
+				// Wait for the API to be loaded.
+				timeout: 10000,
```
Comment (getdave) [inline test/e2e/specs/editor/blocks/navigation.spec.js:49]: Ah right so you're happy with this approach of bumping up the timeout.

I much prefer the approach of not waiting on the implementation details (i.e. network request). Testing user perceivable things is much much better.

**Comment by getdave** [comment]
Ok I'm happy with the structure and intent of this test. Requesting final review(s).

File: test/e2e/specs/editor/blocks/navigation.spec.js (hunk 10 above)
Comment (MaggieCabrera) [reply test/e2e/specs/editor/blocks/navigation.spec.js:49]: oh, interesting, won't this delay the github actions when they run for each commit? I guess it's not too long, but I didn't know this was accepted practice

Comment (getdave) [reply test/e2e/specs/editor/blocks/navigation.spec.js:49]: I guess it will only wait for the block to appear. Which is the same as waiting on the network request.

Comment (MaggieCabrera) [reply test/e2e/specs/editor/blocks/navigation.spec.js:49]: that's smart then, good to know!

**Review by draganescu — APPROVED** [review]
This LGTM! 🚢 it!

File: test/e2e/specs/editor/blocks/navigation.spec.js (hunk 10 above)
Comment (kevin940726) [reply test/e2e/specs/editor/blocks/navigation.spec.js:49]: Yeah, the default timeout is 5 seconds, we're increasing the timeout to give the network requests more time to resolve. It will pass as soon as the page list block is visible.
